janus_core/helpers/descriptors.py

```diff
@@ -0,0 +1,143 @@
+"""Calculate MLIP descriptors for structures."""
+
+from collections.abc import Sequence
+from logging import Logger
+from pathlib import Path
+from typing import Any, Optional
+
+from ase import Atoms
+from ase.io import write
+import numpy as np
+
+from janus_core.helpers.janus_types import ASEWriteArgs, MaybeSequence
+from janus_core.helpers.log import config_logger
+from janus_core.helpers.utils import none_to_dict
+
+
+def calc_descriptors(
+    struct: MaybeSequence[Atoms],
+    struct_name: Optional[str] = None,
+    invariants_only: bool = True,
+    calc_elements: bool = False,
+    write_results: bool = False,
+    write_kwargs: Optional[ASEWriteArgs] = None,
+    log_kwargs: Optional[dict[str, Any]] = None,
+) -> MaybeSequence[Atoms]:
+    """
+    Prepare and call calculation of MLIP descriptors for the given structure(s).
+
+    Parameters
+    ----------
+    struct : MaybeSequence[Atoms]
+        Structure(s) to calculate descriptors for.
+    struct_name : Optional[str]
+        Name of structure. Default is None.
+    invariants_only : bool
+        Whether only the invariant descriptors should be returned. Default is True.
+    calc_elements : bool
+        Whether to calculate mean descriptors for each element. Default is False.
+    write_results : bool
+        True to write out structure with results of calculations. Default is False.
+    write_kwargs : Optional[ASEWriteArgs],
+        Keyword arguments to pass to ase.io.write if saving structure with
+        results of calculations. Default is {}.
+    log_kwargs : Optional[dict[str, Any]]
+        Keyword arguments to pass to `config_logger`. Default is {}.
+
+    Returns
+    -------
+    MaybeSequence[Atoms]
+        Atoms object(s) with descriptors attached as info.
+    """
+    [write_kwargs, log_kwargs] = none_to_dict([write_kwargs, log_kwargs])
+    log_kwargs.setdefault("name", __name__)
+    logger = config_logger(**log_kwargs)
+
+    # Set default name for output file
+    if not struct_name:
+        if isinstance(struct, Sequence):
+            struct_name = struct[0].get_chemical_formula()
+        else:
+            struct_name = struct.get_chemical_formula()
```

Comment:
Do we just want to convert non-`Sequence` structs into a single-element `tuple`?

Reply:
We could do, but then we'd have to convert back at the end, or put in `Atoms` and get back a `tuple`, which might be a bit annoying.